Refactor DFA

z-rui opened the pull request:
- Simplified memory allocation for states
- Removed builder dependency from Dfa class
- Concrete DFA moved to tokengen.cc
- Separated DFA storage and simulation logic
- Reduced copy of strings during simulation

Thoughts?

z-rui committed:
Refactor DFA

- Simplified memory allocation for states
- Removed builder dependency from Dfa class
- Concrete DFA moved to tokengen.cc
- Separated DFA storage and simulation logic
- Reduced copy of strings during simulation

z-rui commented:
Just realized there is another PR that overlaps with this.  Feel free to cherry-pick some ideas from this one and close it.

z-rui committed:
Small fix

Hugozys commented on srcs/dfa.cc:
```diff
@@ -1,109 +1,46 @@
-    _passedAcceptState = nullptr;
-    _passedAcceptStateStr.clear();
-    _buffer.clear();
+Dfa::State Dfa::AddState(Dfa::Action action, bool accepting){
```
Would it be better if this method and those below become part of the interface of the builder class instead of the dfa class?

z-rui replied:
You can make things like ImmutableDFA and MutableDFA (or DFABuiler if you like), and only the latter has non-const methods.

To me, not much difference since you can disable the mutators by passing a const reference.

z-rui committed:
Remove DfaSimulator

Its logic seems very coupled with TokenGen.  It's probably simpler to
absorb the logic into TokenGen.

Since Tokenize() returns a list of tokens from the input string, the
TokenGen class doesn't need to be stateful.

Hugozys reviewed (commented):
Incorporated some of the suggestions into the other PR.

Hugozys commented:
Applied suggestions into the other PR. Close this one. Feel free to reopen if you have other suggestions or ideas.
